src/util-lua-tls.c

```diff
@@ -316,48 +252,112 @@ static int GetCertChain(lua_State *luastate, const Flow *f, int direction)
+{
+    void *state = FlowGetAppState(f);
+    if (state == NULL)
+        return LuaCallbackError(luastate, "error: no app layer state");
+
+    SSLState *ssl_state = (SSLState *)state;
+    SSLStateConnp *connp = NULL;
+
     int direction = LuaStateGetDirection(luastate);
+    if (direction) {
+        connp = &ssl_state->client_connp;
+    } else {
+        connp = &ssl_state->server_connp;
+    }
+    if (connp->cert0_serial == NULL)
+        return LuaCallbackError(luastate, "error: no certificate serial");
 
-    Flow *f = LuaStateGetFlow(luastate);
-    if (f == NULL)
-        return LuaCallbackError(luastate, "internal error: no flow");
+    return LuaPushStringBuffer(
+            luastate, (uint8_t *)connp->cert0_serial, strlen(connp->cert0_serial));
+}
 
-    r = GetCertChain(luastate, f, direction);
+static int LuaTlsGetCertSerial(lua_State *luastate)
+{
+    struct LuaTls *s = (struct LuaTls *)lua_touserdata(luastate, 1);
+    if (s == NULL || s->f == NULL) {
+        LUA_ERROR("failed to get flow");
+    }
 
-    return r;
+    return GetCertSerial(luastate, s->f);
 }
 
-/** \brief register tls lua extensions in a luastate */
-int LuaRegisterTlsFunctions(lua_State *luastate)
+static int GetAgreedVersion(lua_State *luastate, Flow *f)
 {
-    /* registration of the callbacks */
-    lua_pushcfunction(luastate, TlsGetCertNotBefore);
-    lua_setglobal(luastate, "TlsGetCertNotBefore");
+    void *state = FlowGetAppState(f);
+    if (state == NULL)
+        return LuaCallbackError(luastate, "error: no app layer state");
+
+    SSLState *ssl_state = (SSLState *)state;
 
-    lua_pushcfunction(luastate, TlsGetCertNotAfter);
-    lua_setglobal(luastate, "TlsGetCertNotAfter");
+    int direction = LuaStateGetDirection(luastate);
+
+    int version;
+    if (direction) {
+        version = ssl_state->client_connp.version;
+    } else {
+        version = ssl_state->server_connp.version;
+    }
 
-    lua_pushcfunction(luastate, TlsGetVersion);
-    lua_setglobal(luastate, "TlsGetVersion");
+    char ssl_version[SSL_VERSION_MAX_STRLEN];
+    SSLVersionToString(version, ssl_version);
 
-    lua_pushcfunction(luastate, TlsGetCertInfo);
-    lua_setglobal(luastate, "TlsGetCertInfo");
+    lua_pushstring(luastate, (const char *)&ssl_version);
+    return 1;
+}
 
-    lua_pushcfunction(luastate, TlsGetSNI);
-    lua_setglobal(luastate, "TlsGetSNI");
+static int LuaTlsGetVersion(lua_State *luastate)
+{
+    struct LuaTls *s = (struct LuaTls *)lua_touserdata(luastate, 1);
+    if (s == NULL || s->f == NULL) {
+        LUA_ERROR("failed to get flow");
+    }
 
-    lua_pushcfunction(luastate, TlsGetCertSerial);
-    lua_setglobal(luastate, "TlsGetCertSerial");
+    return GetAgreedVersion(luastate, s->f);
+}
 
-    lua_pushcfunction(luastate, TlsGetCertChain);
-    lua_setglobal(luastate, "TlsGetCertChain");
+static const struct luaL_Reg tlslib_meta[] = {
+    // clang-format off
+    { "get_cert_not_before", LuaTlsGetCertNotBefore },
```

Comment:
With other libs we don't hide/imply the directionality. We just get the tx and give the script a way to access the fields. So here it would make sense to do things like `get_client_cert_chain`, `get_server_cert_chain`, etc.

Reply:
I'll rename the functions to incorporate client/server.